### PR TITLE
Add GPU support for model and flows

### PR DESCRIPTION
**Pull request by Baukebrenninkmeijer**
I added GPU support for all forward/inverse passes of the flows that support it and NormalizingFlowModel. The tensors are moved to the same device as the input tensor (so wherever x or z are). For now, the result of sampling will be moved to the CPU, since I expect that to most often be the use-case. Let me know if you have any other suggestions. 

To get it working, put your x, prior and model on the GPU:
```python
device = 'cuda' if torch.cuda.is_available() else 'cpu'
prior = MultivariateNormal(torch.zeros(1).to(device), torch.eye(1).to(device))
model = NormalizingFlowModel(prior, flows).to(device)
x = torch.Tensor(gen_data(args.n)).to(device)
```

From there on, it should work without any problems. 

I didn't change the examples, but can make them use GPU is possible as well. Let me know your preference.

### CODE REVIEW

File: nf/models.py [hunk 1]
```diff
@@ -30,5 +29,5 @@ def inverse(self, z):
     def sample(self, n_samples):
         z = self.prior.sample((n_samples,))
         x, _ = self.inverse(z)
-        return x
+        return x.to('cpu')
```
Comment (tonyduan) [inline nf/models.py]: Don't think sending to CPU here is necessary. We can let the samples stay on the same device as `self.prior`.

File: nf/flows.py [hunk 2]
```diff
@@ -175,8 +175,8 @@ def reset_parameters(self):
         init.uniform_(self.initial_param, -math.sqrt(0.5), math.sqrt(0.5))
 
     def forward(self, x):
-        z = torch.zeros_like(x)
-        log_det = torch.zeros(z.shape[0])
+        z = torch.zeros_like(x).to(x.device)
```
Comment (tonyduan) [inline nf/flows.py]: `torch.zeros_like` should already match the device of its input. 

So no need for `.to(x.device)` on all lines that use `torch.zeros_like`

File: nf/flows.py [hunk 3]
```diff
@@ -175,8 +175,8 @@ def reset_parameters(self):
...
+        log_det = torch.zeros(z.shape[0]).to(x.device)
```
Comment (tonyduan) [inline nf/flows.py]: I believe it's more idiomatic to replace these types of lines with 

`log_det = torch.zeros(z.shape[0], device=x.device)`

**Comment by tonyduan** [comment]
This looks great thanks for adding GPU support. I've left some comments on the PR.

**Review by tonyduan — CHANGES_REQUESTED** [review]
This looks great thanks for adding GPU support. I've left some comments on the PR.

**Comment by Baukebrenninkmeijer** [comment]
Moves device location to tensor initialization instead of after. Removed the cpu as default location for sampling, and is now taking the same location as the prior. Should be good now :).

**Comment by Baukebrenninkmeijer** [comment]
The parameters in OneByOneConv are not registered correctly as parameters, and are not moved to the correct device when we call model.to(device). So now i'm calling the `.to` in the forward and backward call individually.

**Comment by tonyduan** [comment]
Thanks for making the requested changes.

> The parameters in OneByOneConv are not registered correctly as parameters, and are not moved to the correct device when we call model.to(device). So now i'm calling the .to in the forward and backward call individually.

I think we can fix this by replacing the line:

`self.P = torch.tensor(P, dtype = torch.float)`

with:

`self.P = nn.Parameter(torch.tensor(P, dtype = torch.float), requires_grad = False)`

Could you take a stab at this and let me know if this fixes the issue?

**Comment by Baukebrenninkmeijer** [comment]
Yes, i'll have a look when I have time. Hopefully, somewhere later this week.
